File: captchatools-go/capsolver.go
package captchatoolsgo

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"time"
)

type Capsolver struct {
	*Config
}

func (a Capsolver) GetBalance() (float32, error) {
	return a.getBalance()
}

func (c Capsolver) getBalance() (float32, error) {
	// Create payload
	payload := fmt.Sprintf(`{ "clientKey": "%v" }`, c.Api_key)

	// Make POST Request to API
	response := struct {
		ErrorID   int     `json:"errorId"`
		ErrorCode string  `json:"errorCode"`
		Balance   float32 `json:"balance"`
	}{}
	for i := 0; i < 5; i++ {
		resp, err := http.Post(
			"https://api.capsolver.com/getBalance",
			"application/json",
			bytes.NewBufferString(payload),
		)
		if err != nil {
			time.Sleep(1 * time.Second)
			continue
		}

		// Parse Response
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		json.Unmarshal(body, &response)
		if response.ErrorID != 0 {
			return 0, errCodeToError(response.ErrorCode)
		}
		return response.Balance, nil
	}
	return 0, ErrMaxAttempts
}

func (c Capsolver) GetToken(additional ...*AdditionalData) (*CaptchaAnswer, error) {
	return c.getCaptchaAnswer(context.Background(), additional...)
}

func (c Capsolver) GetTokenWithContext(ctx context.Context, additional ...*AdditionalData) (*CaptchaAnswer, error) {
	return c.getCaptchaAnswer(ctx, additional...)
}

// Method to get Queue ID from the API.
func (c Capsolver) getID(data *AdditionalData) (string, error) {
	// Get Payload
	payload, err := c.createPayload(data)
	if err != nil {
		return "", err
	}

	// Make request to get answer
	response := &struct {
		ErrorID   int    `json:"errorId"`
		ErrorCode string `json:"errorCode"`
		TaskID    string `json:"taskId"`
	}{}
	for i := 0; i < 50; i++ {
		resp, err := http.Post("https://api.capsolver.com/createTask", "application/json", bytes.NewBuffer([]byte(payload)))
		if err != nil {
			time.Sleep(3 * time.Second)
			continue
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		json.Unmarshal(body, response)

		// Parse the response
		if response.ErrorID != 0 { // Means there was an error
			return "", errCodeToError(response.ErrorCode)
		}
		return response.TaskID, nil
	}
	return "", ErrMaxAttempts
}

func (c Capsolver) getCaptchaAnswer(ctx context.Context, additional ...*AdditionalData) (*CaptchaAnswer, error) {
	var data *AdditionalData = nil
	if len(additional) > 0 {
		data = additional[0]
	}

	// Get Queue ID
	queueID, err := c.getID(data)
	if err != nil {
		return nil, err
	}

	// Get Captcha Answer
	type Payload struct {
		ClientKey string `json:"clientKey"`
		TaskID    string `json:"taskId"`
	}
	payload, _ := json.Marshal(Payload{
		ClientKey: c.Api_key,
		TaskID:    queueID,
	})
	response := &capmonsterTokenResponse{}
	for i := 0; i < 50; i++ {
		reqToMake, _ := http.NewRequestWithContext(ctx, "POST", "https://api.capsolver.com/getTaskResult", bytes.NewBuffer(payload))
		reqToMake.Header.Add("Content-Type", "application/json")
		resp, err := makeRequest(reqToMake)
		if err != nil {
			if errors.Is(err, context.DeadlineExceeded) {
				return nil, fmt.Errorf("getCaptchaAnswer error: %w", err)
			}
			time.Sleep(3 * time.Second)
			continue
		}

		// Parse Response
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		json.Unmarshal(body, response)

		// Check for any errors
		if response.ErrorID != 0 { // means there was an error
			return nil, errCodeToError(response.ErrorCode)
		}

		// Check if captcha is ready
		if response.Status == "processing" {
			time.Sleep(3 * time.Second)
			continue
		}

		var solution string
		var ua string = response.Solution.UserAgent
		switch c.CaptchaType {
		case V2Captcha, V3Captcha:
			solution = response.Solution.GRecaptchaResponse
		case ImageCaptcha:
			solution = response.Solution.Text
		case CFTurnstile:
			solution = response.Solution.Token
		}

		return newCaptchaAnswer(
			queueID,
			solution,
			c.Api_key,
			c.CaptchaType,
			AnticaptchaSite, // TODO change this
			ua,
		), nil
	}
	return nil, ErrMaxAttempts
}

func (c Capsolver) createPayload(data *AdditionalData) (string, error) {
	type Task struct {
		Type       captchaType `json:"type"`
		WebsiteURL string      `json:"websiteURL"`
		WebsiteKey string      `json:"websiteKey"`

		// Recaptcha V2 Data
		IsInvisible bool   `json:"isInvisible,omitempty"`
		UserAgent   string `json:"userAgent,omitempty"`
		Proxy       string `json:"proxy,omitempty"`

		// Recaptcha V3 Data
		PageAction string  `json:"pageAction,omitempty"`
		MinScore   float32 `json:"minScore,omitempty"`

		// Image Captcha data
		B64Image string `json:"body,omitempty"`
	}
	type Payload struct {
		ClientKey string `json:"clientKey"`
		Task      Task   `json:"task"`
	}

	p := Payload{
		ClientKey: c.Api_key,
		Task: Task{
			Type:       c.CaptchaType,
			WebsiteURL: c.CaptchaURL,
			WebsiteKey: c.Sitekey,
		},
	}

	switch c.CaptchaType {
	case ImageCaptcha:
		p.Task.Type = "ImageToTextTask"
		if data != nil {
			p.Task.B64Image = data.B64Img
		}
	case V2Captcha:
		p.Task.Type = "ReCaptchaV2TaskProxyLess"
		p.Task.IsInvisible = c.IsInvisibleCaptcha
		if data != nil && data.Proxy != nil {
			p.Task.Type = "ReCaptchaV2Task"
		}
	case V3Captcha:
		p.Task.Type = "ReCaptchaV3TaskProxyLess"
		p.Task.MinScore = c.MinScore
		p.Task.PageAction = c.Action
		if data != nil && data.Proxy != nil {
			p.Task.Type = "ReCaptchaV3Task"
		}
	case HCaptcha:
		p.Task.Type = "HCaptchaTaskProxyLess"
		if data != nil && data.Proxy != nil {
			p.Task.Type = "HCaptchaTurboTask"
		}

	}

	// Check for any additional data about the task
	if data != nil && c.CaptchaType != ImageCaptcha {
		if data.UserAgent != "" {
			p.Task.UserAgent = data.UserAgent
		}
		if data.Proxy != nil {
			p.Task.Proxy = data.Proxy.StringFormatted()
		}
	}

	// TODO add softkey id

	// Return
	encoded, _ := json.Marshal(p)
	return string(encoded), nil
}

File: captchatools/capsolver.py
from .errors import ErrCodeToException
from . import Harvester
import time
import requests
from typing import Optional

BASEURL = "https://api.capsolver.com"



class Capsolver(Harvester):
    '''
    This object will contain the data to interact with Capsolver API
    '''
    def get_balance(self) -> float:
        payload = {"clientKey": self.api_key}
        for _ in range(5):
            try:
                resp = requests.post(BASEURL + "/getBalance", json=payload ,timeout=20).json()
                if resp["errorId"] == 1: # Means there was an error
                    ErrCodeToException("Capsolver", resp["errorCode"])
                return float(resp["balance"])
            except requests.RequestException:
                pass

    def get_token(
            self, 
            b64_img: Optional[str] = None, 
            user_agent: Optional[str] = None, 
            proxy: Optional[str] = None, 
            proxy_type: Optional[str] = "HTTP",
            rq_data: Optional[str] = None
        ):
        # Get ID
        task_id, already_solved = self.__get_id(
            b64_img=b64_img,
            user_agent=user_agent,
            proxy=proxy,
            proxy_type=proxy_type,
            rq_data=rq_data
        )

        # Check if token was already retrieved
        if already_solved:
            return task_id
        
        # Get Answer
        return self.__get_answer(task_id)
    
    def __create_payload(self, **kwargs):
        payload = {"clientKey":self.api_key,"task":{}}

        # Add data based on the captcha type
        if self.captcha_type == "image" or self.captcha_type == "normal":
            payload["task"]["type"] = "ImageToTextTask"
            payload["task"]["body"] = kwargs.get("b64_img", "")
        elif self.captcha_type == "v2":
            payload["task"]["type"] = "ReCaptchaV2TaskProxyLess"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
            if self.invisible_captcha:
                payload["task"]["isInvisible"] = self.invisible_captcha
            if kwargs.get("proxy", None) is not None:
                payload["task"]["type"] = "ReCaptchaV2Task"
        elif self.captcha_type == "v3":
            payload["task"]["type"] = "RecaptchaV3TaskProxyless"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
            payload["task"]["minScore"] = self.min_score
            payload["task"]["pageAction"] = self.action
            if kwargs.get("proxy", None) is not None:
                payload["task"]["type"] = "ReCaptchaV3Task"
        elif self.captcha_type == "hcap" or self.captcha_type == "hcaptcha":
            payload["task"]["type"] = "HCaptchaTaskProxyless"
            if kwargs.get("proxy", None) is not None:
                payload["task"]["type"] = "HCaptchaTask"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
        elif self.captcha_type == "hcaptchaturbo":
            payload["task"]["type"] = "HCaptchaTurboTask"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
        elif self.captcha_type == "fun" or self.captcha_type == "funcaptcha":
            payload["task"]["type"] = "FunCaptchaTaskProxyless"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
            if kwargs.get("proxy", None) is not None:
                payload["task"]["type"] = "FunCaptchaTask"

        # Add Global Data
        if self.soft_id is not None:
            payload["appId"] = self.soft_id
        if kwargs.get("proxy", None) is not None:
            splitted = kwargs.get("proxy").split(":")
            payload["task"]["proxyAddress"] = splitted[0]
            try:
                payload["task"]["proxyPort"] = int(splitted[1])
            except Exception:
                payload["task"]["proxyPort"] = splitted[1]
            if len(splitted) >=4:
                payload["task"]["proxyLogin"] = splitted[2]
                payload["task"]["proxyPassword"] = splitted[3]
            payload["task"]["proxyType"] = kwargs.get("proxy_type", "http")
        if kwargs.get("user_agent", None) is not None:
            payload["task"]["userAgent"] = kwargs.get("user_agent")
        if kwargs.get("rq_data", None) is not None:
            payload["task"]["enterprisePayload"] = {"rqdata":  kwargs.get("rq_data")}
        return payload
    
    def __get_id(self,**kwargs):
        '''
        This method gets a task ID for a captcha token.

        A tuple is returned from this method ( Optional[task_id|captcha_token] | bool)
        If the second index is True, that means a captcha token is returned instead of a task ID
        '''
        # Create Payload
        payload = self.__create_payload(**kwargs)
        
        # Get token & return it
        for _ in range(50):
            try:
                resp = requests.post(BASEURL + "/createTask" , json=payload, timeout=20).json()
                if resp["errorId"] != 0: # Means there was an error:
                    ErrCodeToException("Capsolver", resp["errorCode"])
                
                # Check if there is an answer already available
                if resp["status"] == "ready":
                    if resp["solution"].get("text", None) is not None:
                        answer = resp["solution"].get("text", None)
                    elif resp["solution"].get("gRecaptchaResponse", None) is not None:
                        answer = resp["solution"].get("gRecaptchaResponse", None)
                    return (answer, True)
                return (resp["taskId"], False)
            except (requests.RequestException, KeyError):
                pass
    
    def __get_answer(self,task_id:int):
        payload = {"clientKey":self.api_key,"taskId": task_id}
        for _ in range(100):
            try:
                response = requests.post(BASEURL + "/getTaskResult",json=payload,timeout=20,).json()
                if response["errorId"] != 0: # Means there was an error
                    ErrCodeToException("Capsolver", response["errorDescription"])
                if response["status"] == "processing":
                    time.sleep(3)
                    continue
                if self.captcha_type == "normal" or self.captcha_type == "image":
                    return response["solution"]["text"]
                else:
                    return response["solution"]["gRecaptchaResponse"]
            except (requests.RequestException, KeyError):
                pass    
